docs/adr/0031-event-list-synchronization-in-chat-scroll-view.md

```diff
@@ -0,0 +1,133 @@
+    return thisTransactionId == otherTransactionId;
+  }
+
+  // Cross-match: event's transaction_id with other's eventId
+  if (thisTransactionId != null && thisTransactionId == other.eventId) return true;
+  if (otherTransactionId != null && otherTransactionId == eventId) return true;
+
+  return false;
+}
+```
+
+### 2. Hash Map Optimization
+
+Created `buildEventMap` in `EventListExtension` ([lib/utils/matrix_sdk_extensions/event_list_extension.dart](lib/utils/matrix_sdk_extensions/event_list_extension.dart)) that builds event maps with dual keys (eventId and transactionId) for O(1) lookups:
+
+```dart
+static Map<String, Event> buildEventMap(List<Event> events) {
+  final map = <String, Event>{};
+  for (final event in events) {
+    map[event.eventId] = event;
+    final transactionId = event.unsigned?['transaction_id'] as String?;
+    if (transactionId != null) {
+      map[transactionId] = event;
+    }
+  }
+  return map;
+}
+```
+
+### 3. Reusable Event List Extension
+
+Created `EventListExtension` with `syncEventLists` method that handles:
+- Removing deleted events from display lists
+- Updating changed events (e.g., sending → sent transitions)
+- Adding new events at appropriate positions
+- Preventing duplicate global key exceptions
+- Achieving O(n) time complexity for all operations
+
+### 4. Build-Time Safety Check
+
+Added safety check in ChatScrollView build method to skip rendering events not found in `widget.events`:
+
+```dart
+final currentEventIndex = widget.events.indexWhere(
+  (e) => _isSameEvent(e, currentEvent),
+);
+
+if (currentEventIndex == -1) {
+  return const SizedBox.shrink(); // Event was deleted
+}
+```
+
+### 5. Comprehensive Testing
+
+Created [test/utils/matrix_sdk_extensions/event_list_extension_test.dart](test/utils/matrix_sdk_extensions/event_list_extension_test.dart) with 18 unit tests covering:
+- Event matching by eventId and transaction_id
+- Hash map building with dual keys
+- Event lookup in maps
+- Event synchronization scenarios (removal, addition, updates, duplicates)
+
+## Consequences
+
+### Positive
+
+1. **Performance**: Reduced time complexity from O(m×n) to O(n) using hash maps
+2. **Reliability**: Eliminated duplicate global key exceptions
+3. **Correctness**: Deleted events are properly removed from display
+4. **Maintainability**: Simplified ChatScrollView `didUpdateWidget` from ~160 lines to ~30 lines
+5. **Code Quality**: Extracted reusable logic to `EventListExtension`
+6. **Test Coverage**: 18 comprehensive unit tests ensuring correctness
+7. **Reusability**: Event synchronization logic can be used in other widgets
+
+### Negative
+
+1. Memory overhead from maintaining event maps (minimal impact)
+2. Additional layer of abstraction to understand
+
+### Known Considerations
+
+- **Matrix SDK Event Loops**: The Matrix SDK can sometimes add/remove the same event in rapid succession, creating a loop. The build-time safety check ensures such events are properly hidden when removed.
```

Comment:
Is it still the case (duplicate event) with the updated matrix sdk? 

If yes then, maybe we should fix it upstream no? Is it really at the app level we should fix this kind of issues ?